Improve user defined StringContext error messages

When working with a custom StringContext or adding custom methods on scala.StringContext, type mismatch errors are for the transformed/generated trees/code and so they are one step removed from the source code.  Examples:

```scala
object H1 {
   case class StringContext(i: Int)

   val x = 3
   val str = s"a: $x"
}
object H2 {
   case class StringContext(i: Int*)

   val x = 3
   val str = s"a: $x"
}
object H3 {
   case class StringContext(strs: String*) {
      def s: Int = 3
   }

   val x = 3
   val str = s"a: $x"
}
```

```
H.scala:5: error: too many arguments for method apply: (i: Int)H1.StringContext in object StringContext
   val str = s"a: $x"
                  ^
H.scala:11: error: type mismatch;
 found   : String("a: ")
 required: Int
   val str = s"a: $x"
                  ^
H.scala:19: error: Int does not take parameters
   val str = s"a: $x"
                  ^
three errors found
```

--- a/src/compiler/scala/tools/nsc/ast/parser/Parsers.scala
+++ b/src/compiler/scala/tools/nsc/ast/parser/Parsers.scala
@@ -1328,8 +1328,10 @@ self =>
      */
     def literal(isNegated: Boolean = false, inPattern: Boolean = false, start: Offset = in.offset): Tree = atPos(start) {
       def finish(value: Any): Tree = try newLiteral(value) finally in.nextToken()
-      if (in.token == INTERPOLATIONID)
-        interpolatedString(inPattern = inPattern)
+      if (in.token == INTERPOLATIONID) {
+        if (inPattern) interpolatedString(inPattern)
+        else withPlaceholders(interpolatedString(inPattern), isAny = true) // interpolator params are Any* by definition
+      }
       else if (in.token == SYMBOLLIT) {
         def msg(what: String) =
           s"""symbol literal is $what; use Symbol("${in.strVal}") instead"""
@@ -1382,41 +1384,32 @@ self =>
     }
 
     private def interpolatedString(inPattern: Boolean): Tree = {
-      def errpolation() = syntaxErrorOrIncompleteAnd("error in interpolated string: identifier or block expected",
-                                                     skipIt = true)(EmptyTree)
-      // Like Swiss cheese, with holes
-      def stringCheese: Tree = atPos(in.offset) {
-        val start        = in.offset
-        val interpolator = in.name.encoded // ident() for INTERPOLATIONID
-
-        val partsBuf = new ListBuffer[Tree]
-        val exprsBuf = new ListBuffer[Tree]
-        in.nextToken()
-        while (in.token == STRINGPART) {
-          partsBuf += literal()
-          exprsBuf += (
-            if (inPattern) dropAnyBraces(pattern())
-            else in.token match {
-              case IDENTIFIER => atPos(in.offset)(Ident(ident()))
-              //case USCORE   => freshPlaceholder()  // ifonly etapolation
-              case LBRACE     => expr()              // dropAnyBraces(expr0(Local))
-              case THIS       => in.nextToken(); atPos(in.offset)(This(tpnme.EMPTY))
-              case _          => errpolation()
-            }
-          )
-        }
-        if (in.token == STRINGLIT) partsBuf += literal()
+      val start        = in.offset
+      val interpolator = in.name.encoded // ident() for INTERPOLATIONID
+      val partsBuf     = ListBuffer.empty[Tree]
+      val exprsBuf     = ListBuffer.empty[Tree]
+      in.nextToken()
+      while (in.token == STRINGPART) {
+        partsBuf += literal()
+        exprsBuf += (
+          if (inPattern) dropAnyBraces(pattern())
+          else in.token match {
+            case IDENTIFIER => atPos(in.offset)(Ident(ident()))
+            case LBRACE     => expr()
+            case THIS       => in.nextToken(); atPos(in.offset)(This(tpnme.EMPTY))
+            case _          => syntaxErrorOrIncompleteAnd("error in interpolated string: identifier or block expected", skipIt = true)(EmptyTree)
+          }
+        )
+      }
+      if (in.token == STRINGLIT) partsBuf += literal()
 
       // Documenting that it is intentional that the ident is not rooted for purposes of virtualization
       //val t1 = atPos(o2p(start)) { Select(Select (Ident(nme.ROOTPKG), nme.scala_), nme.StringContextName) }
-        val t1 = atPos(o2p(start)) { Ident(nme.StringContextName) }
-        val t2 = atPos(start) { Apply(t1, partsBuf.toList) }
-        t2 setPos t2.pos.makeTransparent
-        val t3 = Select(t2, interpolator) setPos t2.pos
-        atPos(start) { Apply(t3, exprsBuf.toList) }
-      }
-      if (inPattern) stringCheese
-      else withPlaceholders(stringCheese, isAny = true) // string interpolator params are Any* by definition
+      val t1 = atPos(o2p(start)) { Ident(nme.StringContextName) }
+      val t2 = atPos(start) { Apply(t1, partsBuf.toList) } updateAttachment InterpolatedString
+      t2 setPos t2.pos.makeTransparent
+      val t3 = Select(t2, interpolator) setPos t2.pos
+      atPos(start) { Apply(t3, exprsBuf.toList) } updateAttachment InterpolatedString
     }
 
 /* ------------- NEW LINES ------------------------------------------------- */

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -4938,11 +4938,11 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             loop(tree)
           }
           val retry = typeErrors.forall(_.errPos != null) && (errorInResult(fun) || errorInResult(tree) || args.exists(errorInResult))
-          typingStack.printTyping({
-            val funStr = ptTree(fun) + " and " + (args map ptTree mkString ", ")
-            if (retry) "second try: " + funStr
-            else "no second try: " + funStr + " because error not in result: " + typeErrors.head.errPos+"!="+tree.pos
-          })
+          typingStack.printTyping {
+            val funStr = s"${ptTree(fun)} and ${args.map(ptTree).mkString(", ")}"
+            if (retry) s"second try: $funStr"
+            else s"no second try: $funStr because error not in result: ${typeErrors.head.errPos}!=${tree.pos}"
+          }
           if (retry) {
             val Select(qual, name) = fun
             tryTypedArgs(args, forArgMode(fun, mode)) match {
@@ -4957,7 +4957,18 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
               case _ => ()
             }
           }
-          typeErrors foreach context.issue
+          def adjust(err: AbsTypeError) =
+            if (tree.hasAttachment[InterpolatedString.type])
+              tree match {
+                case Apply(sc @ Ident(nme.StringContextName), _) =>
+                  if (sc.isErroneous) err
+                  else NormalTypeError(tree, s"${err.errMsg}; signature for interpolation must be `StringContext.apply(String*)`")
+                case Apply(Select(_, nm), badargs) =>
+                  if (badargs.exists(_.isErroneous)) err
+                  else NormalTypeError(tree, s"${err.errMsg}; incompatible interpolation method $nm")
+              }
+            else err
+          typeErrors.foreach(err => context.issue(adjust(err)))
           warnings foreach { case (p, m, c, s) => context.warning(p, m, c, s) }
           setError(treeCopy.Apply(tree, fun, args))
         }

--- a/src/reflect/scala/reflect/internal/StdAttachments.scala
+++ b/src/reflect/scala/reflect/internal/StdAttachments.scala
@@ -129,4 +129,6 @@ trait StdAttachments {
 
   // When typing a Def with this attachment, change the owner of its RHS from origalOwner to the symbol of the Def
   case class ChangeOwnerAttachment(originalOwner: Symbol)
+
+  case object InterpolatedString extends PlainAttachment
 }

--- a/src/reflect/scala/reflect/runtime/JavaUniverseForce.scala
+++ b/src/reflect/scala/reflect/runtime/JavaUniverseForce.scala
@@ -67,6 +67,7 @@ trait JavaUniverseForce { self: runtime.JavaUniverse  =>
     this.MultiargInfixAttachment
     this.NullaryOverrideAdapted
     this.ChangeOwnerAttachment
+    this.InterpolatedString
     this.noPrint
     this.typeDebug
     // inaccessible: this.posAssigner

--- a/test/files/neg/t6352.check
+++ b/test/files/neg/t6352.check
@@ -1,0 +1,23 @@
+t6352.scala:9: error: too many arguments (found 2, expected 1) for method apply: (i: Int): H1.StringContext in object StringContext; signature for interpolation must be `StringContext.apply(String*)`
+  val str = s"a: $x"
+            ^
+t6352.scala:15: error: type mismatch;
+ found   : String("a: ")
+ required: Int; signature for interpolation must be `StringContext.apply(String*)`
+  val str = s"a: $x"
+            ^
+t6352.scala:23: error: Int does not take parameters; incompatible interpolation method s
+  val str = s"a: $x"
+            ^
+t6352.scala:31: error: not found: value y
+  val bad = s"b: $y"
+                  ^
+t6352.scala:38: error: value t is not a member of X2.StringContext
+  val bad = t"a: $x"
+            ^
+t6352.scala:45: error: type mismatch;
+ found   : Int
+ required: String; incompatible interpolation method s
+  val bad = s"a: $x"
+            ^
+6 errors

--- a/test/files/neg/t6352.scala
+++ b/test/files/neg/t6352.scala
@@ -1,0 +1,46 @@
+//scalac: -Yimports:java.lang
+
+import scala.Int
+
+object H1 {
+  case class StringContext(i: Int)
+
+  val x = 3
+  val str = s"a: $x"
+}
+object H2 {
+  case class StringContext(i: Int*)
+
+  val x = 3
+  val str = s"a: $x"
+}
+object H3 {
+  case class StringContext(strs: String*) {
+    def s: Int = 3
+  }
+
+  val x = 3
+  val str = s"a: $x"
+}
+object X1 {
+  case class StringContext(strs: String*) {
+    def s(args: Int*): String = strs.mkString + args.sum
+  }
+  val x = 3
+  val str = s"a: $x"
+  val bad = s"b: $y"
+}
+object X2 {
+  case class StringContext(strs: String*) {
+    def s(args: Int*): String = strs.mkString + args.sum
+  }
+  val x = 3
+  val bad = t"a: $x"
+}
+object X3 {
+  case class StringContext(strs: String*) {
+    def s(args: String*): String = scala.StringContext(strs: _*).s(args: _*)
+  }
+  val x = 3
+  val bad = s"a: $x"
+}

